[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20084 [JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.**
* #20083 [JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.
* #20082 [JIT] SubgraphRewriter: Rename pattern fusion to subgraph rewrite.

Previously, we could only replace a match with a single node. With this
cahnge, we will be able to specify an arbitrary subgraph to insert
instead of the match. A special case would be an empty replacement
subgraph, which would essentially mean that we just want to remove
matches from the original graph.

Differential Revision: [D15190191](https://our.internmc.facebook.com/intern/diff/D15190191)

Commit:
[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.

Previously, we could only replace a match with a single node. With this
cahnge, we will be able to specify an arbitrary subgraph to insert
instead of the match. A special case would be an empty replacement
subgraph, which would essentially mean that we just want to remove
matches from the original graph.

raghuramank100 commented on torch/csrc/jit/passes/subgraph_rewrite.h:
```diff
@@ -58,21 +58,14 @@ class TORCH_API SubgraphRewriter {
   /** \brief Register a custom rewrite pattern.
    *
    * The method takes four parameters specifying the pattern:
```
Nit: This function only takes two inputs

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1457,7 +1457,10 @@ def test_pattern_based_rewrite(self):
 graph(%a, %b, %c):
   %q = aten::mul(%a, %b)
```
Can we have a test for the special case outlined in the comments? What about the corner case when we want to replace the entire graph with an empty graph?

ZolotukhinM replied:
> Can we have a test for the special case outlined in the comments? 

What case are you referring to? We currently have tests for: 1) fusion (replace many with one), 2) combining (replace many with many), 3) removal (replace something with nothing).

> What about the corner case when we want to replace the entire graph with an empty graph?

We would be able to do that if we add support for wildcards. Currently there is no way to do it (but I also have a question, why do you want do this:) ).

raghuramank100 replied:
I agree, this is only a corner case. I think I missed seeing the test case for removal of nodes, so that is sufficient.

zdevito commented on torch/csrc/jit/passes/subgraph_rewrite.cpp:
```diff
@@ -44,6 +43,39 @@ void SubgraphRewriter::runOnGraph(std::shared_ptr<Graph>& graph) {
+ * incoming values for the new subgraph. Return a vector of values produced by
+ * the inserted subgraph.
+ */
+static std::vector<Value*> insertSubgraphCloneAfter(
```
This appears to be an implementation of ir.h's `inlineCallTo` which is the preferred way of inlining a call of one graph into another.

zdevito reviewed (approved):
This looks good, assuming use of inlineCallTo

Commit:
Update on "[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter."

[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.

Previously, we could only replace a match with a single node. With this
cahnge, we will be able to specify an arbitrary subgraph to insert
instead of the match. A special case would be an empty replacement
subgraph, which would essentially mean that we just want to remove
matches from the original graph.

gh-metadata: pytorch pytorch 20084 gh/ZolotukhinM/82/head

Commit:
Update on "[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter."

[JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.

Previously, we could only replace a match with a single node. With this
cahnge, we will be able to specify an arbitrary subgraph to insert
instead of the match. A special case would be an empty replacement
subgraph, which would essentially mean that we just want to remove
matches from the original graph.

gh-metadata: pytorch pytorch 20084 gh/ZolotukhinM/82/head
